### PR TITLE
[SYCL][DOC] Add extension for FPGA kernel interface properties

### PR DESCRIPTION
**Pull request by tiwaria1**
## SYCL extension contains the following new kernel properties
- `streaming_interface<...>`
- `register_map_interface<...>`
The first two properties take enum arguments that provide the compiler information about whether the logic downstream to the kernel will back-pressure the kernel or not.
- `pipelined<N>`
Takes an integer, non-zero values specify minimum cycles between kernel invocations, and 0 specifies that pipelining should be disabled.

### CODE REVIEW

File: sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+
+- link:sycl_ext_oneapi_properties.asciidoc[sycl_ext_oneapi_properties]
+- link:sycl_ext_oneapi_kernel_properties.asciidoc[sycl_ext_oneapi_kernel_properties]
+
+== Feature Test Macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification section 6.3.3 "Feature test macros".  Therefore, an
+implementation supporting this extension must predefine the macro
+`SYCL_EXT_ONEAPI_KERNEL_PROPERTIES` to one of the values defined in the table
```
Comment (tiwaria1) [inline sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc:66]:
```suggestion
`SYCL_EXT_INTEL_FPGA_KERNEL_INTERFACE_PROPERTIES` to one of the values defined in the table
```

Comment (GarveyJoe) [inline sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc:66]:
```suggestion
`SYCL_EXT_INTEL_FPGA_KERNEL_INTERFACE_PROPERTIES` to one of the values defined in the table
```

File: sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+
+[%header,cols="1,5"]
+|===
+|Value |Description
+|1     |Initial extension version.  Base features are supported.
+|===
+
+== Overview
+
+This extension specifies the kernel properties designed for kernels that target
```
Comment (GarveyJoe) [inline sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc:79]: This sentence is simultaneously too specific (as the "the" implies that no other extensions might add FPGA kernel properties) and too vague (there's no indication of what these properties are about).  I suggest something like this instead:

This extension introduces kernel properties to specify how or when control and data signals can be passed in to or out of an FPGA kernel.  These properties are meaningless on non-FPGA devices and can be ignored on such devices.

Given that there are only two properties, you could even go further and summarize them both like so:

The ip_interface_streaming property requires the kernel to have a dedicated output port for each control or data signal, as opposed to using a shared CSR interface.  The ip_pipelined property directs the compiler to pipeline calls to the kernel so that multiple invocations of the kernel can be in flight simultaneously.

File: sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+ The property takes a constant boolean template argument. When the value of the
+ argument is `true` it adds the requirement that the kernel module should not
+ take a 'stall' control flow signal from downstream logic as input. The kernel
+ can take such an input signal otherwise. Pass the value `true` to this argument
+ when the downstream logic is known to not stall.
+ 
+ For convenience, a constant bool named `stall_free_return` with the value
+ `true` is provided.
+
+|`ip_pipelined`
```
Comment (GarveyJoe) [inline sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc:132]: I'm not a big fan of the "ip" at the beginning of each of these names.  What concept are you trying to get across with this prefix?

File: sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+[%header,cols="1,5"]
+|===
+|Value |Description
+|1     |Initial extension version.  Base features are supported.
+|===
+
+== Overview
+
+This extension specifies the kernel properties designed for kernels that target
+the FPGA device. These properties are supported by the Intel FPGA backend
```
Comment (GarveyJoe) [inline sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc:80]: Which compilers support this extension is irrelevant for the sake of the definition of the extension itself.

File: sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+} // namespace ext
+} // namespace sycl
+```
+
+|===
+|Property|Description
+|`ip_interface_streaming`
+|The `ip_interface_streaming` property adds the requirement that the kernel must
+ have dedicated ports for input / output signals in the generated
+ Register Transfer Level (RTL) module.
```
Comment (GarveyJoe) [inline sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc:121]: I would drop the mention of RTL.  It confuses things to refer to the description of the hardware as being separate from the hardware itself.  
Also, I think your intent is for this to apply to both control and data signals.  Can you state the explicitly?  And assuming that this should apply to data signals as well, should it apply to all inputs/outputs from the kernel, i.e. pipes as well, or only kernel arguments?

File: sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+ 
+ The property takes a constant boolean template argument. When the value of the
+ argument is `true` it adds the requirement that the kernel module should not
+ take a 'stall' control flow signal from downstream logic as input. The kernel
```
Comment (GarveyJoe) [inline sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc:125]: Why is the stall-freedom a template parameter of ip_interface_streaming instead of its own property?  How can a user guarantee that done signals are not lost in a pipelined kernel with a CSR interface if there is no way to backpressure the done signal?

File: sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+
+== Introduction
+IMPORTANT: This specification is a draft.
+
+NOTE: Khronos(R) is a registered trademark and SYCL(TM) and SPIR(TM) are
+trademarks of The Khronos Group Inc.  OpenCL(TM) is a trademark of Apple Inc.
+used by permission by Khronos.
+
+This extension introduces a replacement for the kernel attributes defined in
+Section 5.8.1 of the SYCL 2020 specification, in the form of a property list
```
[the remark this answers is not in the document]
Comment (tiwaria1) [reply sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc]: Updated.

File: sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc (hunk 2 above)
Comment (tiwaria1) [reply sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc]: Updated. Since it's 4 properties now, I have placed the additional description in the properties table.

File: sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc (hunk 3 above)
Comment (tiwaria1) [reply sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc]: Changed the names to:
```
streaming_kernel
csr_kernel
stall_free_return_kernel
pipelined_kernel
```
1. Perhaps `stall_free_return_kernel` should be renamed to `no_backpressure_kernel`
2. Other kernel properties do not have a '_kernel' post-fix. I think we should instead use the names:
```
streaming_interface
csr_interface
no_backpressure_interface
pipelined
```
Thoughts?

File: sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc (hunk 4 above)
Comment (tiwaria1) [reply sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc]: Updated.

File: sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc (hunk 5 above)
Comment (tiwaria1) [reply sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc]: Added explicit statement. Specified that it should apply to kernel arguments and control flow signals only. Including pipes seems to make sense. I need to think a bit more about including them.

File: sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc (hunk 6 above)
Comment (tiwaria1) [reply sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc]: It should be its own property. Updated.

File: sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc [hunk 8]
```diff
@@ -0,0 +1,198 @@
...
+
+Joe Garvey, Intel +
+Abhishek Tiwari, Intel
+
+== Dependencies
+
+This extension is written against the SYCL 2020 specification, Revision 4 and
+the following extensions:
+
+- link:sycl_ext_oneapi_properties.asciidoc[sycl_ext_oneapi_properties]
```
[the remark this answers is not in the document]
Comment (tiwaria1) [reply sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc]: I removed the hyperlink for now. Need to figure out how to link to a file in a different directory.

File: sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc (hunk 3 above)
Comment (GarveyJoe) [reply sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc]: #2: agreed.  Let's be consistent with other kernel properties.
#1: let's run it by internal users to see whether they prefer "stall_free" or "no_backpressure"

File: sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc [hunk 9]
```diff
@@ -0,0 +1,213 @@
...
+```c++
+struct KernelFunctor {
+
+  KernelFunctor(int* a, int* b, int* c) : a(a), b(b), c(c) {}
+
+  void operator()() const {
+    *a = *b + *c;
+  }
+
+  auto get(sycl::ext::oneapi::experimental::properties_tag) {
```
Comment (keryell) [inline sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc:183]: Is this a magic function that encode the properties? How it is connected to the `single_task` below?

File: sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc [hunk 10]
```diff
@@ -0,0 +1,213 @@
...
+ Control and Status Register (CSR) map.
+
+|`stall_free_return_kernel`
+|The `stall_free_return_kernel` property provides the compiler the guarantee
+ that downstream logic will not back-pressure the kernel.
+
+|`pipelined_kernel`
+| The `pipelined_kernel` property directs the compiler to pipeline calls to the
+ kernel so that multiple invocations of the kernel can be in flight
+ simultaneously.
```
Comment (keryell) [inline sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc:144]: Do you mean several instances of a `single_task` or this is pipelining at the work-item level?

File: sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc [hunk 11]
```diff
@@ -0,0 +1,213 @@
...
+
+The example assumes that the kernel will not be back-pressured by the downstream
+logic and hence uses the property `stall_free_return_kernel`:
+
+```c++
+sycl::ext::oneapi::experimental::properties properties{
+  sycl::ext::oneapi::experimental::streaming_kernel,
+  sycl::ext::oneapi::experimental::stall_free_return_kernel};
+
+q.single_task(properties, [=]() {
```
Comment (keryell) [inline sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc:164]:
```suggestion
q.single_task(properties, [=] {
```

File: sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc (hunk 3 above)
Comment (tiwaria1) [reply sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc]: Updated based on the discussion with internal users.

File: sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc (hunk 8 above)
[the remark this answers is not in the document]
Comment (tiwaria1) [reply sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc]: Corrected the link.

File: sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc [hunk 12]
```diff
@@ -0,0 +1,296 @@
...
+ The following properties have been provided for convenience:
+ `streaming_interface_accept_downstream_stall`,
+ `streaming_interface_remove_downstream_stall`.
+
+|`register_map_interface`
+|The `register_map_interface` property adds the requirement that the kernel must have
+ its input / output control and kernel argument data signals placed in a shared
+ Control and Status Register (CSR) map. The following values are supported:
+
+ * `wait_for_done_read`: Provides the compiler the guarantee that downstream
```
Comment (GarveyJoe) [inline sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc:187]: I thought we didn't want to use a read-on-clear approach and had instead discussed requiring the user to overwrite the done bit with a 0 if they want backpressure.  If we go that way, we should probably call this "wait_for_done_write" or "wait_for_done_clear".

Comment (GarveyJoe) [reply sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc:187]: Also, this definition is vague and misleading.  The user is guaranteeing anything to the compiler, they are asking us to build logic that backpressures the kernel until they've told us they've seen the done.

File: sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc [hunk 13]
```diff
@@ -0,0 +1,296 @@
...
+    logic will back-pressure the kernel.
+
+ * `do_not_wait_for_done_read`: Provides the compiler the guarantee that
```
Comment (GarveyJoe) [inline sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc:190]: This one is more of a guarantee but it still seems like the wrong choice of words.  The user is telling the compiler that it can build hardware that doesn't backpressure the kernel.  I.e. that the user's system is either tolerant to dropped dones or that they are going to throttle the kernel themselves sufficiently to ensure that no done are dropped.

File: sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc [hunk 14]
```diff
@@ -0,0 +1,296 @@
...
+```
+
+|===
+|Property|Description
+|`streaming_interface`
+|The `streaming_interface` property adds the requirement that the kernel must
+ have dedicated ports for input / output signals. This applies for both
+ control, and kernel argument data signals. The following values are supported:
+
+ * `accept_downstream_stall`: Provides the compiler the guarantee that
```
Comment (GarveyJoe) [inline sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc:168]: A "guarantee that downstream logic will back-pressure the kernel" sounds like a guarantee that the ready signal is never high and thus that the kernel never needs to do anything.  That's definitely not what you intended.  You could reword this as: "Instructs the compiler to build a backpressure input port to the kernel (e.g. a ready signal for the avalon-ST protocol)."

File: sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc [hunk 15]
```diff
@@ -0,0 +1,296 @@
...
+    downstream logic will back-pressure the kernel.
+
+ * `remove_downstream_stall`: Provides the compiler the guarantee that
```
Comment (GarveyJoe) [inline sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc:171]: It goes beyond that, right?  This directs the compiler to omit the port entirely.

File: sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc [hunk 16]
```diff
@@ -0,0 +1,296 @@
...
+ `register_map_interface_wait_for_done_read`,
+ `register_map_interface_do_not_wait_for_done_read`.
+
+|`pipelined`
+|Two sets of values are accepted as property parameters - a boolean, and another
+ property `initiation_interval`.
+
+ When the boolean parameter is set to `true`, the property directs the compiler
+ to pipeline calls to the kernel such that multiple invocations of the kernel
+ can be in flight simultaneously. When the boolean paramter is set to `false`,
```
Comment (GarveyJoe) [inline sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc:207]:
```suggestion
 can be in flight simultaneously. When the boolean parameter is set to `false`,
```

File: sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc [hunk 17]
```diff
@@ -0,0 +1,296 @@
...
+ the compiler will not pipeline kernel invocations.
+
+ The second property argument should be used to specify how many cycles can
```
Comment (GarveyJoe) [inline sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc:210]:
```suggestion
 The second property argument specifies how many cycles can
```

File: sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc [hunk 18]
```diff
@@ -0,0 +1,296 @@
...
+ exist between multiple invocations of the kernel. The `initiation_interval`
```
Comment (GarveyJoe) [inline sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc:211]:
```suggestion
 exist between successive invocations of the kernel. The `initiation_interval`
```

File: sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc [hunk 19]
```diff
@@ -0,0 +1,296 @@
...
+Device compilers that do not support this extension may accept and ignore these
+ properties.
+
+== Adding a Property List to a Kernel Launch
+
+A simple example of using this extension to launch a kernel with a streaming
+ interface is shown below.
+
+The example assumes that the kernel will not be back-pressured by the downstream
+logic and hence uses the property `stall_free_return_kernel`:
```
Comment (GarveyJoe) [inline sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc:234]: This is the old name and also the example below does accept a stall so these need to be made consistent.

File: sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc [hunk 20]
```diff
@@ -0,0 +1,296 @@
...
+sycl::ext::oneapi::experimental::properties properties{
+  sycl::ext::oneapi::experimental::streaming_interface_accept_downstream_stall};
+
+q.single_task(properties, [=] {
+  *a = *b + *c;
+}).wait();
+```
+
+The example below shows how to launch a pipelined kernel with a streaming
+interface, and with every kernel invocation being launched every 2 cycles.
```
Comment (GarveyJoe) [inline sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc:246]:
```suggestion
interface, and with a new kernel invocation being launched every 2 cycles.
```

File: sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc (hunk 10 above)
Comment (GarveyJoe) [reply sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc]: This refers to several instances of a kernel, most commonly a single_task kernel.  It does not refer to pipelining of work-items.  However, there's no reason this couldn't be applied to a parallel_for kernel to request that successive launches of that kernel be pipelined _in addition to_ the usual work item pipelining.  In such a kernel, the work items could be injected into the datapath one after another and, after the last work item from a given kernel invocation has been injected, the first work-item of the next invocation could be injected without waiting for the pipeline to drain.  

I don't see much value in a control to enable or disable pipelining at the work-item level because that's nearly always beneficial.  This invocation pipelining is more situational though as it only provides value if the user intends to launch the same kernel multiple times.

File: sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc [hunk 21]
```diff
@@ -0,0 +1,296 @@
...
+data signals can be passed in to or out of an FPGA kernel. These properties are
+meaningless on non-FPGA devices and can be ignored on such devices.
+
+== Kernel Properties
+
+```c++
+namespace sycl {
+namespace ext {
+namespace oneapi {
+namespace experimental {
```
Comment (keryell) [inline sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc:89]: C++17 is good for FPGA :-)
```suggestion
namespace sycl::ext::oneapi::experimental {
```
Along this, it should be good to have at least one example showing how a namespace alias like
```c++
namespace oe = sycl::ext::oneapi::experimental;
```
can be used to write shorter code, just in case some HLS people without any SYCL or C++ knowledge can see how to write compact code and not think that SYCL is overly verbose.

**Comment by tiwaria1** [comment]
ping @intel/dpcpp-specification-reviewers

File: sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc [hunk 22]
```diff
@@ -0,0 +1,302 @@
...
+ is equivalent to one of the values listed above, but the choice is
+ implementation defined.
+
+ The following properties have been provided for convenience:
+ `register_map_interface_wait_for_done_write`,
+ `register_map_interface_do_not_wait_for_done_write`.
+
+|`pipelined`
+|Two sets of values are accepted as property parameters - a boolean, and another
+ property `initiation_interval`.
```
Comment (steffenlarsen) [inline sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc:202]: Is there a benefit to defining `initiation_interval` as a property? I worry that having it use the property interface may wrongfully give the user the idea that it is to be used like a property, while it is simply intended for use inside the `pipelined` property. It also adds more to the implementation, as properties need to define a number of things under the hood, whereas if we define it as some simple wrapper class we can avoid that.

Comment (tiwaria1) [reply sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc]: Hi Steffen, I have updated the spec, please review again. Thank you!

File: sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc [hunk 23]
```diff
@@ -96,10 +96,9 @@ enum class register_map_interface_options_enum {
   do_not_wait_for_done_write
 };
 
-struct initiation_interval_key {
-  template <int initiation_interval_value>
-  using value_t = property_value<initiation_interval_key,
-    std::integral_constant<int, initiation_interval_value>>;
+template <int N>
+struct initiation_interval {
+  int initiation_interval_value = N;
```
Comment (steffenlarsen) [inline sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc:101]: Is this expected to mutate? Otherwise it would be better to have it be `static constexpr`. Also, the naming of the value member seems a little verbose, given it simply repeats the type name. Would `value` be too simple a name?

Comment (tiwaria1) [reply sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc]: Value is not expected to mutate so I will make it a `static constexpr` and yes I can rename the field to `value` as I don't think that the more verbose name is needed.

**Review by steffenlarsen — APPROVED** [review]
LGTM!

File: sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc [hunk 24]
```diff
@@ -0,0 +1,296 @@
...
+This is a proposed extension specification, intended to gather community
+feedback. Interfaces defined in this specification may not be implemented yet
+or may be in a preliminary state. The specification itself may also change in
+incompatible ways before it is finalized. *Shipping software products should not
+rely on APIs defined in this specification.*
+
+== Overview
+
+This extension introduces kernel properties to specify how or when control and
+data signals can be passed in to or out of an FPGA kernel. These properties are
```
Comment (GarveyJoe) [inline sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc:57]:
```suggestion
data signals can be passed into or out of an FPGA kernel. These properties are
```

File: sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc [hunk 25]
```diff
@@ -0,0 +1,296 @@
...
+|`register_map_interface`
+|The `register_map_interface` property adds the requirement that the kernel must
+ have its input / output control and kernel argument data signals placed in a
+ shared Control and Status Register (CSR) map. The following values are
+ supported:
+
+ * `wait_for_done_write`: Directs the compiler to generate logic that
+    back-pressures the kernel until the kernel is notified that its completion
+    has been detected. The kernel will be notified when the register it writes
+    its completion signal to, is set to 1.
```
Comment (GarveyJoe) [inline sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc:181]:
```suggestion
    its completion signal to is set to 1.
```

File: sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc [hunk 26]
```diff
@@ -0,0 +1,296 @@
...
+ the compiler will not pipeline kernel invocations.
+
+ The second argument specifies how many cycles can exist between successive
+ invocations of the kernel:
+
+ * `true, initiation_interval<N>`: The compiler will pipeline multiple kernel
+   invocations such that an invocation is launched every N cycles.
+ 
+ If the `initiation_interval` argument is not specified, the compiler will
+ attempt to launch the kernel invocations as fast as it can.
```
Comment (GarveyJoe) [inline sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc:211]: I propose we leave this as an implementation detail.  As written it implies that an implementation must aggressively pursue II=1 even if that doesn't make sense (e.g. because the kernel contains a loop with II=1 with high trip count which will be the bottleneck anyways).

Comment (tiwaria1) [reply sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc]: Updated to mark this as implementation defined.

**Review by keryell — COMMENTED** [review]
`pipelined` could be simplified to take a single parameter with value `false`, `true` or some II value.
The current cartesian product with 2 parameters is useless since `false` and an II is meaningless.

**Comment by keryell** [comment]
Also having a default `true` value for the `pipelined` when used seems an interesting shortcut.

File: sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc [hunk 27]
```diff
@@ -0,0 +1,296 @@
...
+|`register_map_interface`
+|The `register_map_interface` property adds the requirement that the kernel must
+ have its input / output control and kernel argument data signals placed in a
+ shared Control and Status Register (CSR) map. The following values are
+ supported:
+
+ * `wait_for_done_write`: Directs the compiler to generate logic that
+    back-pressures the kernel until the kernel is notified that its completion
+    has been detected. The kernel will be notified when the register it writes
+    its completion signal to is set to 1.
```
Comment (GarveyJoe) [inline sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc:181]:
```suggestion
    its completion signal to is set to 0.
```

If it's a "done" register then the kernel should write 1 to it when it finishes and the user should write 0 to it to say "I've seen your done".

File: sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc [hunk 28]
```diff
@@ -0,0 +1,296 @@
...
+
+ The second argument specifies the minimum number of cycles between successive
+ invocations of the kernel:
+
+ * `true, initiation_interval<N>`: The compiler will pipeline multiple kernel
+   invocations such that an invocation can be launched every N cycles if one is available.
+ 
+ If the `initiation_interval` argument is not specified, the behavior is
+ implementation defined.
+
```
Comment (GarveyJoe) [inline sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc:212]: You might want to mention here that if the bool is false, the initiation_interval parameter is meaningless.

File: sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc [hunk 29]
```diff
@@ -0,0 +1,296 @@
...
+Joe Garvey, Intel +
+Abhishek Tiwari, Intel
+
+== Dependencies
+
+This extension is written against the SYCL 2020 specification, Revision 4 and
+the following extensions:
+
+- link:../experimental/sycl_ext_oneapi_properties.asciidoc[sycl_ext_oneapi_properties]
+- link:sycl_ext_oneapi_kernel_properties.asciidoc[sycl_ext_oneapi_kernel_properties]
```
Comment (GarveyJoe) [inline sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc:44]: Links are broken

**Comment by tiwaria1** [comment]
@steffenlarsen I addressed the remaining comments. Please review and merge this PR. Based on your namespace comment on a different PR I have updated the namespace here to be `sycl::ext::intel::experimental` since this is a FPGA specific extension.

File: sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc (hunk 29 above)
Comment (tiwaria1) [reply sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc:44]: They seem to be working now.
oneapi properties is in experimental folder: https://github.com/intel/llvm/tree/sycl/sycl/doc/extensions/experimental
and oneapi kernel properties will be in the proposed folder after 5755 PR goes in : https://github.com/intel/llvm/pull/5755/files

File: sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc [hunk 30]
```diff
@@ -0,0 +1,287 @@
...
+
+struct streaming_interface_key {
+  template <streaming_interface_options_enum option>
+  using value_t = property_value<streaming_interface_key,
+    std::integral_constant<streaming_interface_options_enum, option>>;
+};
+
+struct register_map_interface_key {
+  template <register_map_interface_options_enum option>
+  using value_t = property_value<register_map_interface_key,
```
Comment (tiwaria1) [inline sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc:107]: Hi @steffenlarsen , Instead of `property_value`, should this be rewritten as `sycl::ext::oneapi::properties::property_value` ? Or perhaps I should add a `using sycl::ext::oneapi::properties` at the top?

File: sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc [hunk 31]
```diff
@@ -0,0 +1,287 @@
...
+
+The example assumes that the kernel will not accept a signal that can
+back-pressure it and hence uses the property
+`streaming_interface_remove_downstream_stall`:
+
+```c++
+using sycl::ext::intel::experimental;
+{
+  ...
+  properties kernel_properties{streaming_interface_remove_downstream_stall};
```
Comment (steffenlarsen) [inline sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc:226]:
```suggestion
  auto kernel_properties = properties{streaming_interface_remove_downstream_stall};
```
And similar in other places.

**Review by steffenlarsen — COMMENTED** [review]
Just a few more comments, then I think I'm happy!

File: sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc (hunk 31 above)
Comment (keryell) [reply sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc:226]: This is bike-shedding but I usually push into the other direction, when `auto` is useless and longer. :-)

Comment (steffenlarsen) [reply sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc:226]: Honestly, I wasn't sure it would accept the type without the template arguments explicitly there, but I guess I've learned something new. In that case I definitely like the old version better. Thanks @keryell! 😄

File: sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc (hunk 30 above)
Comment (tiwaria1) [reply sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc]: Updated!

**Comment by tiwaria1** [comment]
> Just a few more comments, then I think I'm happy!

Hey Steffen, I addressed the other comments. Please take a look.

**Review by steffenlarsen — APPROVED** [review]
LGTM!
